src/Common/src/System/Globalization/IdnMapping.Unix.cs

```diff
@@ -11,44 +11,39 @@ private unsafe string GetAsciiCore(char* unicode, int count)
             uint flags = Flags;
             CheckInvalidIdnCharacters(unicode, count, flags, nameof(unicode));
 
-            const int StackAllocThreshold = 512;
-            if (count < StackAllocThreshold)
-            {
-                char* output = stackalloc char[count];
-                return GetAsciiCore(unicode, count, flags, output, count, reattempt: true);
-            }
-            else
+            // Each unicode character is represented by 2 ASCII chars
+            // and a "xn--" prefix of length 4
+            const int StackallocThreshold = 512;
+            int estimatedLength = (int)Math.Min(count * 2L + 4, StackallocThreshold);
+            int actualLength;
+            if (estimatedLength < StackallocThreshold)
             {
-                char[] output = new char[count];
-                fixed (char* pOutput = output)
+                char* outputStack = stackalloc char[estimatedLength];
+                actualLength = Interop.GlobalizationNative.ToAscii(flags, unicode, count, outputStack, estimatedLength);
+                if (actualLength > 0 && actualLength <= estimatedLength)
```

Comment:
I just tried this locally, and there are a bunch of cases in our tests where the actualLength is coming back longer than the estimatedLength.  Is that expected?
```
Estimated: 12, Actual: 18, Str: à.א̈
Estimated: 12, Actual: 18, Str: À.א̈
Estimated: 12, Actual: 18, Str: à̈.א
Estimated: 12, Actual: 18, Str: À̈.א
Estimated: 10, Actual: 15, Str: ü.ü
Estimated: 10, Actual: 15, Str: Ü.Ü
Estimated: 10, Actual: 15, Str: Ü.ü
Estimated: 16, Actual: 17, Str: 日本語.jp
Estimated: 16, Actual: 17, Str: 日本語.JP
Estimated: 16, Actual: 17, Str: 日本語.Jp
Estimated: 6, Actual: 7, Str: ☕
Estimated: 6, Actual: 7, Str: 夙
Estimated: 8, Actual: 9, Str: את
Estimated: 10, Actual: 11, Str: א׳ת
Estimated: 10, Actual: 11, Str: א7ת
Estimated: 10, Actual: 11, Str: א٧ת
Estimated: 8, Actual: 9, Str: ஹ்
Estimated: 6, Actual: 7, Str: ஹ
Estimated: 12, Actual: 13, Str: لٰۭۯ
Estimated: 10, Actual: 11, Str: لٰۯ
Estimated: 10, Actual: 12, Str: لۭۯ
Estimated: 8, Actual: 10, Str: لۯ
Estimated: 10, Actual: 11, Str: لٰۭ
Estimated: 6, Actual: 7, Str: ل
Estimated: 6, Actual: 7, Str: ٻ
Estimated: 10, Actual: 12, Str: ݰ𐹸
Estimated: 6, Actual: 7, Str: ā
Estimated: 6, Actual: 7, Str: ā
Estimated: 6, Actual: 7, Str: ā
Estimated: 18, Actual: 19, Str: そのスピードで
Estimated: 18, Actual: 19, Str: そのスピードで
Estimated: 18, Actual: 19, Str: そのスピードで
Estimated: 38, Actual: 41, Str: そのスピードで.パフィーdeルンバ
Estimated: 38, Actual: 41, Str: そのスピードで.パフィーdeルンバ
Estimated: 18, Actual: 19, Str: そのスピードで
```